File: sdk/python/aragora/namespaces/partner.py
"""
Partner API Namespace.

Provides methods for partner management:
- Partner registration and profile
- API key management
- Usage statistics
- Webhook configuration
- Rate limit information

Endpoints:
    POST   /api/v1/partners/register     - Register as partner
    GET    /api/v1/partners/me           - Get partner profile
    POST   /api/v1/partners/keys         - Create API key
    GET    /api/v1/partners/keys         - List API keys
    DELETE /api/v1/partners/keys/{id}    - Revoke API key
    GET    /api/v1/partners/usage        - Get usage statistics
    POST   /api/v1/partners/webhooks     - Configure webhook
    GET    /api/v1/partners/limits       - Get rate limits
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class PartnerAPI:
    """
    Synchronous Partner API.

    Provides methods for partner registration, API key management,
    and usage tracking.

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> # Register as a partner
        >>> result = client.partner.register(
        ...     name="My Company",
        ...     email="partner@example.com",
        ...     company="My Company Inc",
        ... )
        >>> # Create an API key
        >>> key = client.partner.create_api_key(
        ...     name="Production Key",
        ...     scopes=["debates:read", "debates:write"],
        ... )
        >>> print(f"API Key: {key['key']}")  # Only shown once!
    """

    def __init__(self, client: AragoraClient) -> None:
        self._client = client

    # =========================================================================
    # Registration
    # =========================================================================

    def register(
        self,
        name: str,
        email: str,
        company: str | None = None,
    ) -> dict[str, Any]:
        """
        Register as a partner.

        Args:
            name: Partner name.
            email: Contact email.
            company: Optional company name.

        Returns:
            Dict with partner_id, status, referral_code.
        """
        data: dict[str, Any] = {"name": name, "email": email}
        if company:
            data["company"] = company

        return self._client.request("POST", "/api/v1/partners/register", json=data)

    def get_profile(self) -> dict[str, Any]:
        """
        Get current partner profile.

        Requires X-Partner-ID header.

        Returns:
            Dict with partner stats and profile.
        """
        return self._client.request("GET", "/api/v1/partners/me")

    # =========================================================================
    # API Keys
    # =========================================================================

    def create_api_key(
        self,
        name: str = "API Key",
        scopes: list[str] | None = None,
        expires_in_days: int | None = None,
    ) -> dict[str, Any]:
        """
        Create a new API key.

        Args:
            name: Key name.
            scopes: Optional list of scopes.
            expires_in_days: Optional expiration days.

        Returns:
            Dict with key_id, key (only shown once!), scopes.
        """
        data: dict[str, Any] = {"name": name}
        if scopes:
            data["scopes"] = scopes
        if expires_in_days:
            data["expires_in_days"] = expires_in_days

        return self._client.request("POST", "/api/v1/partners/keys", json=data)

    def list_api_keys(self) -> dict[str, Any]:
        """
        List all API keys.

        Returns:
            Dict with keys array and counts.
        """
        return self._client.request("GET", "/api/v1/partners/keys")

    def revoke_api_key(self, key_id: str) -> dict[str, Any]:
        """
        Revoke an API key.

        Args:
            key_id: The key ID to revoke.

        Returns:
            Dict with success message.
        """
        return self._client.request("DELETE", f"/api/v1/partners/keys/{key_id}")

    # =========================================================================
    # Usage
    # =========================================================================

    def get_usage(self, days: int = 30) -> dict[str, Any]:
        """
        Get usage statistics.

        Args:
            days: Number of days to look back (1-365).

        Returns:
            Dict with usage statistics.
        """
        return self._client.request("GET", "/api/v1/partners/usage", params={"days": days})

    # =========================================================================
    # Webhooks
    # =========================================================================

    def configure_webhook(self, url: str) -> dict[str, Any]:
        """
        Configure webhook endpoint.

        Args:
            url: Webhook URL (must be HTTPS).

        Returns:
            Dict with webhook_url and webhook_secret (only shown once!).
        """
        return self._client.request("POST", "/api/v1/partners/webhooks", json={"url": url})

    # =========================================================================
    # Limits
    # =========================================================================

    def get_limits(self) -> dict[str, Any]:
        """
        Get rate limits for partner tier.

        Returns:
            Dict with tier, limits, current usage, allowed status.
        """
        return self._client.request("GET", "/api/v1/partners/limits")


class AsyncPartnerAPI:
    """Asynchronous Partner API."""

    def __init__(self, client: AragoraAsyncClient) -> None:
        self._client = client

    # =========================================================================
    # Registration
    # =========================================================================

    async def register(
        self,
        name: str,
        email: str,
        company: str | None = None,
    ) -> dict[str, Any]:
        """Register as a partner."""
        data: dict[str, Any] = {"name": name, "email": email}
        if company:
            data["company"] = company

        return await self._client.request("POST", "/api/v1/partners/register", json=data)

    async def get_profile(self) -> dict[str, Any]:
        """Get current partner profile."""
        return await self._client.request("GET", "/api/v1/partners/me")

    # =========================================================================
    # API Keys
    # =========================================================================

    async def create_api_key(
        self,
        name: str = "API Key",
        scopes: list[str] | None = None,
        expires_in_days: int | None = None,
    ) -> dict[str, Any]:
        """Create a new API key."""
        data: dict[str, Any] = {"name": name}
        if scopes:
            data["scopes"] = scopes
        if expires_in_days:
            data["expires_in_days"] = expires_in_days

        return await self._client.request("POST", "/api/v1/partners/keys", json=data)

    async def list_api_keys(self) -> dict[str, Any]:
        """List all API keys."""
        return await self._client.request("GET", "/api/v1/partners/keys")

    async def revoke_api_key(self, key_id: str) -> dict[str, Any]:
        """Revoke an API key."""
        return await self._client.request("DELETE", f"/api/v1/partners/keys/{key_id}")

    # =========================================================================
    # Usage
    # =========================================================================

    async def get_usage(self, days: int = 30) -> dict[str, Any]:
        """Get usage statistics."""
        return await self._client.request("GET", "/api/v1/partners/usage", params={"days": days})

    # =========================================================================
    # Webhooks
    # =========================================================================

    async def configure_webhook(self, url: str) -> dict[str, Any]:
        """Configure webhook endpoint."""
        return await self._client.request("POST", "/api/v1/partners/webhooks", json={"url": url})

    # =========================================================================
    # Limits
    # =========================================================================

    async def get_limits(self) -> dict[str, Any]:
        """Get rate limits for partner tier."""
        return await self._client.request("GET", "/api/v1/partners/limits")

File: sdk/typescript/src/namespaces/partner.ts
/**
 * Partner API Namespace
 *
 * Provides methods for partner management:
 * - Partner registration and profile
 * - API key management
 * - Usage statistics
 * - Webhook configuration
 * - Rate limit information
 *
 * Endpoints:
 *   POST   /api/v1/partners/register     - Register as partner
 *   GET    /api/v1/partners/me           - Get partner profile
 *   POST   /api/v1/partners/keys         - Create API key
 *   GET    /api/v1/partners/keys         - List API keys
 *   DELETE /api/v1/partners/keys/{id}    - Revoke API key
 *   GET    /api/v1/partners/usage        - Get usage statistics
 *   POST   /api/v1/partners/webhooks     - Configure webhook
 *   GET    /api/v1/partners/limits       - Get rate limits
 */

/**
 * Partner profile information.
 */
export interface PartnerProfile {
  partner_id: string;
  name: string;
  email: string;
  company?: string;
  status: 'pending' | 'active' | 'suspended';
  tier: 'free' | 'basic' | 'professional' | 'enterprise';
  referral_code: string;
  created_at: string;
  updated_at?: string;
  total_api_calls: number;
  active_keys: number;
}

/**
 * API key information.
 */
export interface PartnerApiKey {
  key_id: string;
  name: string;
  prefix: string;
  scopes: string[];
  created_at: string;
  expires_at?: string;
  last_used_at?: string;
  is_active: boolean;
}

/**
 * Usage statistics.
 */
export interface PartnerUsage {
  period_start: string;
  period_end: string;
  total_requests: number;
  requests_by_endpoint: Record<string, number>;
  requests_by_day: Array<{ date: string; count: number }>;
  average_latency_ms: number;
  error_rate: number;
}

/**
 * Rate limit information.
 */
export interface PartnerLimits {
  tier: string;
  limits: {
    requests_per_minute: number;
    requests_per_day: number;
    max_concurrent: number;
    max_batch_size: number;
  };
  current_usage: {
    requests_this_minute: number;
    requests_today: number;
    concurrent_requests: number;
  };
  allowed: boolean;
  reset_at?: string;
}

/**
 * Options for partner registration.
 */
export interface RegisterPartnerOptions {
  /** Partner name */
  name: string;
  /** Contact email */
  email: string;
  /** Optional company name */
  company?: string;
}

/**
 * Options for creating an API key.
 */
export interface CreateApiKeyOptions {
  /** Key name (default "API Key") */
  name?: string;
  /** Optional list of scopes */
  scopes?: string[];
  /** Optional expiration in days */
  expiresInDays?: number;
}

/**
 * Client interface for making HTTP requests.
 */
interface PartnerClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * Partner API namespace.
 *
 * Provides methods for partner registration, API key management,
 * and usage tracking.
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // Register as a partner
 * const result = await client.partner.register({
 *   name: 'My Company',
 *   email: 'partner@example.com',
 *   company: 'My Company Inc',
 * });
 *
 * // Get partner profile
 * const profile = await client.partner.getProfile();
 * console.log(`Tier: ${profile.tier}`);
 *
 * // Create an API key
 * const key = await client.partner.createApiKey({
 *   name: 'Production Key',
 *   scopes: ['debates:read', 'debates:write'],
 * });
 * console.log(`API Key: ${key.key}`); // Only shown once!
 *
 * // Check rate limits
 * const limits = await client.partner.getLimits();
 * console.log(`Requests today: ${limits.current_usage.requests_today}`);
 * ```
 */
export class PartnerAPI {
  constructor(private client: PartnerClientInterface) {}

  // =========================================================================
  // Registration
  // =========================================================================

  /**
   * Register as a partner.
   *
   * @param options - Registration options
   * @returns Partner ID, status, and referral code
   */
  async register(options: RegisterPartnerOptions): Promise<{
    partner_id: string;
    status: string;
    referral_code: string;
  }> {
    const data: Record<string, unknown> = {
      name: options.name,
      email: options.email,
    };
    if (options.company) data.company = options.company;

    return this.client.request('POST', '/api/v1/partners/register', {
      json: data,
    });
  }

  /**
   * Get current partner profile.
   *
   * Requires X-Partner-ID header.
   *
   * @returns Partner profile with stats
   */
  async getProfile(): Promise<PartnerProfile> {
    return this.client.request('GET', '/api/v1/partners/me');
  }

  // =========================================================================
  // API Keys
  // =========================================================================

  /**
   * Create a new API key.
   *
   * @param options - API key creation options
   * @returns Key information including the key itself (only shown once!)
   */
  async createApiKey(options?: CreateApiKeyOptions): Promise<{
    key_id: string;
    key: string;
    scopes: string[];
  }> {
    const data: Record<string, unknown> = {
      name: options?.name ?? 'API Key',
    };
    if (options?.scopes) data.scopes = options.scopes;
    if (options?.expiresInDays) data.expires_in_days = options.expiresInDays;

    return this.client.request('GET', '/api/v1/partners/keys', {
      json: data,
    });
  }

  /**
   * List all API keys.
   *
   * @returns List of API keys with counts
   */
  async listApiKeys(): Promise<{
    keys: PartnerApiKey[];
    total: number;
    active: number;
  }> {
    return this.client.request('GET', '/api/v1/partners/keys');
  }

  /**
   * Revoke an API key.
   *
   * @param keyId - The key ID to revoke
   * @returns Success message
   */
  async revokeApiKey(keyId: string): Promise<{
    success: boolean;
    message: string;
  }> {
    return this.client.request('DELETE', `/api/v1/partners/keys/${keyId}`);
  }

  // =========================================================================
  // Usage
  // =========================================================================

  /**
   * Get usage statistics.
   *
   * @param days - Number of days to look back (1-365, default 30)
   * @returns Usage statistics
   */
  async getUsage(days = 30): Promise<PartnerUsage> {
    return this.client.request('GET', '/api/v1/partners/usage', {
      params: { days },
    });
  }

  // =========================================================================
  // Webhooks
  // =========================================================================

  /**
   * Configure webhook endpoint.
   *
   * @param url - Webhook URL (must be HTTPS)
   * @returns Webhook URL and secret (only shown once!)
   */
  async configureWebhook(url: string): Promise<{
    webhook_url: string;
    webhook_secret: string;
  }> {
    return this.client.request('GET', '/api/v1/partners/webhooks', {
      json: { url },
    });
  }

  // =========================================================================
  // Limits
  // =========================================================================

  /**
   * Get rate limits for partner tier.
   *
   * @returns Rate limit information with current usage
   */
  async getLimits(): Promise<PartnerLimits> {
    return this.client.request('GET', '/api/v1/partners/limits');
  }
}
